laurent22/joplin
# Desktop: Fixes #9122: Fix underscores escaped within some text-only URLs

# Summary

This pull request updates Turndown to only escape `_`s if they are preceded by punctuation, a space-like character, or the start of a line.

This seems to be sufficient to prevent MarkdownIt from creating italicized regions based on these characters.

Fixes #9122.

# Manual testing

This pull request has an automated test. However, if desired, it can be tested manually with the following note:
````markdown
Some URLs in the Rich_Text_Editor contain `_` characters, but haven't been converted to links yet. For example, https://www.example.com/a_test_of_links, or www.example.com/a_test_of_links.

We should preserve the underscores \_without escaping them_ to prevent the links from breaking.

This should also correctly handle unicode characters. For example, punctuation❯\_requires escapes_, but 𝔏𝔈𝔗𝔗𝔈𝕽_𝔠𝔥𝔞𝔯𝔞𝔠𝔱𝔢𝔯𝔰_and_897_numbers_𝒟on_'t.

\_Note_ that what \[\_causes_\] a \`\_\` to create italics_ seems to depend only on the character before and an escape at the \_beginning_ seems to be sufficient.

\_s also don't need escapes if _ followed _ by a \_space..

``a_ ` _test``

```
This
1. is
2. a
3. test

_of_ escapes! \test
```
````

Pasting the above into the markdown editor, switching to the rich text editor and editing, undoing the edit, then switching back to the markdown editor, should result in the same markdown.

# Notes

- There may be edge cases that this pull request does not handle correctly.
- This pull request includes a commit that updates Rollup &mdash; the version Turndown used didn't recognize Unicode `\p` escapes:

        [!] (commonjs plugin) SyntaxError: Invalid regular expression: /(^|\p{Punctuation}|\p{Separator})_/: Invalid escape (20:4) in /home/usernamehere/Documents/joplin/packages/turndown/src/turndown.js

- A pull request fixing a previous version of this issue, https://github.com/laurent22/joplin/pull/8873 only fixed the case where URLs had already been converted to HTML links. Its change has not been reverted as #8873 is not specific to underscore escapes (and should handle edge cases that this PR may not).
